docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+
+When the compiler is in doubt, an annotation will nudge it towards relying on type inference, while an assertion will force it to accept the user-supplied type.
+
+<!-- TODO: Add example -->
+
+##### When instantiating an empty container type, provide a type annotation
+
+This is one case where type inference is unable to reach a useful conclusion without user-provided information. Since the compiler cannot arbitrarily restrict the range of types that could be inserted into the container, it has to assume the widest type, which is often `any`. It's up to the user to narrow that into the intended type by adding an explicit annotation.
+
+🚫
+
+```typescript
+const tokens = []; // Type 'any[]'
+const tokensMap = new Map(); // Type 'Map<any, any>'
+```
+
+✅
+
+```typescript
+const tokens: string[] = []; // Type 'string[]'
+const tokensMap = new Map<string, Token>(); // Type 'Map<string, Token>'
+```
+
+### Type Assertions
+
+`as` assertions are inherently unsafe. They overwrite type-checked and inferred types with user-supplied types that suppress compiler errors.
+
+They should only be introduced into the code if the accurate type is unreachable through other means.
+
+##### Document safe or necessary use of type assertions
+
+When a type assertion is absolutely necessary due to constraints or is even safe due to runtime checks, we should document the reasoning behind its usage in the form of a comment.
+
+<!-- TODO: Add example -->
+
+#### Avoid `as`
+
+Type assertions make the code brittle against changes.
+
+While TypeScript and ESLint will flag some unsafe, structurally unsound, or redundant type assertions, they will generally accept user-supplied types without further type-checking.
+
+This can cause silent failures or false negatives where errors are suppressed. This is especially damaging as the codebase accumulates changes over time. Type assertions may continue to silence errors, even though the type itself, or the type's relationship to the rest of the code may have been altered so that the asserted type is no longer valid.
+
+##### Redundant or unnecessary `as` assertions are not flagged for removal
+
+Type assertions can also cause false positives, because assertions are independent expressions, untied to the type errors they were intended to fix. Even if code drift fixes or removes a particular type error, the type assertions that were put in place to fix that error will provide no indication that they are no longer necessary and now should be removed.
+
+```typescript
+enum Direction {
+  Up = 'up',
+  Down = 'down',
+  Left = 'left',
+  Right = 'right',
+}
+const directions = Object.values(Direction);
+
+// Error: Element implicitly has an 'any' type because index expression is not of type 'number'.(7015)
```

Comment:
I'm having trouble following this example. What is this error meant to represent - is this the type error that we should see here?

When I try this in the TypeScript Playground, I get this:

```
Conversion of type 'string[]' to type 'keyof directions[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.(2352)
```

Reply:
Oh I see, this error was present but was eliminated by the type assertions. Got it. Maybe we could add some context here to explain that.

Also, I'll leave a separate comment for the error I saw, I may have a solution

Reply:
Applied here: https://github.com/MetaMask/contributor-docs/pull/74/commits/5d81b66240fc5f299e943da9afb8bea27a4a51f1